app/controllers/auth/SelfAssessmentAction.scala

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2021 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers.auth
+
+import com.google.inject.ImplementedBy
+import config.ApplicationConfig
+import play.api.mvc.Results.Redirect
+import play.api.mvc.{ActionRefiner, Result}
+import services.{CitizenDetailsService, SucccessMatchingDetailsResponse}
+import uk.gov.hmrc.auth.core.ConfidenceLevel
+import uk.gov.hmrc.domain.SaUtr
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.play.HeaderCarrierConverter
+import javax.inject.Inject
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class SelfAssessmentActionImpl @Inject()(
+  citizenDetailsService: CitizenDetailsService,
+  ninoAuthAction: NinoAuthAction,
+  appConfig: ApplicationConfig)(implicit ec: ExecutionContext)
+    extends SelfAssessmentAction {
+
+  override protected def refine[A](
+    request: AuthenticatedRequest[A]): Future[Either[Result, AuthenticatedRequest[A]]] = {
+
+    implicit val hc = HeaderCarrierConverter.fromHeadersAndSession(request.headers, Some(request.session))
+
+    if (request.saUtr.isEmpty) {
+      ninoAuthAction.getNino().flatMap { atsNinoResponse =>
```

Comment:
Think we still need to cover the IR-SA-AGENT scenario here, so they can come in if their saUTR is empty as long as they are an active agent (see the `authAction` logic

Reply:
would it be enough to check for agentRef in the request or does it need the check for active enrolment like this one?             val isAgentActive: Boolean = enrolments.find(_.key == "IR-SA-AGENT").map(_.isActivated).getOrElse(false)
I'm not sure how to get the enrolments in the refine here

Reply:
Will need to do the latter, you can pass through this information from the `AuthenticatedRequest`